src/iter.rs

```diff
@@ -1,108 +1,158 @@
-use core::slice;
-use core::convert::TryInto;
-use core::convert::TryFrom;
-
 #[allow(missing_docs)]
 pub struct Bytes<'a> {
-    slice: &'a [u8],
-    pos: usize
+    start: *const u8,
+    end: *const u8,
+    cursor: *const u8,
+    phantom: core::marker::PhantomData<&'a ()>,
 }
 
 #[allow(missing_docs)]
 impl<'a> Bytes<'a> {
     #[inline]
     pub fn new(slice: &'a [u8]) -> Bytes<'a> {
+        let start = slice.as_ptr();
+        let end = unsafe { start.add(slice.len()) };
+        let cursor = start;
         Bytes {
-            slice,
-            pos: 0
+            start,
+            end,
+            cursor,
+            phantom: core::marker::PhantomData,
         }
     }
 
     #[inline]
     pub fn pos(&self) -> usize {
-        self.pos
+        self.cursor as usize - self.start as usize
```

Comment:
This and the other arithmetic below in `len` don't end up costing more than just reading the integer? Or perhaps the other operations cost less, and they balance out?

Reply:
This PR does enable perf improvements (just not without other changes in `lib.rs`  (and `simd/*`) to match the more optimal usage patterns this unblocks). On `req/req` (M1 Air) this takes us from `~114ns/iter` to `~70ns/iter` which is rather significant.

`.pos()` is infrequently read, it's not in the hotpath and #139 mostly stops using it because it's somewhat quirky.

Similarly `.len()` isn't required or optimal for bounds checking, for example in SIMD batches `.len()` mutates with each loop as you shorten the buffer. So whilst a slice `.len()` is just a read, it's crucial to notice that the actual slice (and thus `.len()`) we're checking is constantly changing since `bytes.as_ref()` is computing a new slice

What you really want is just not to overrun the buffer and perform an OOB read, so what you can do is compute `max(start, end - SIMD_BATCH_SIZE)` and bounds check that the current `ptr` is less than that. That `endsimd_ptr` is constant across an entire simd batch and broadly an entire parse call so the compiler can compute it once at top of loop keep it in a register whereas `.len` will constantly mutate (with every single `.advance()` call)

The core thing to understand with this PR is the following: `(*const u8, *const u8)` and `(*const u8, usize)` can both equally denote a slice or contiguous chunk of memory. In our case we're modelling a `Cursor` over that `slice`, let's look at those structs and their invariants as they mutate over time:

```rs
// This PR
struct PtrCursor {
  start: *const u8, // Increases monotonically
  cursor: *const u8, // Increases monotonically 
  end: *const, // Constant, never changes post initialization
}

// master
struct SliceCursor {
  slice_start: *const u8, // Increases monotonicallyx
  slice_len: usize, // Decreases monotonically
  pos: usize, // Increases, then is reset 
}
```

Conceptually `PtrCursor` is much easier to reason about (and optimize), especially when you note that it's basically:
```
struct PtrCursorFull {
    buf_start: *const u8,
    buf_end: *const u8,
}
```